File: URI/Paradigmas/1288 - Canhao de Destruicao/canhao.py
# -*- coding: utf-8 -*-
# Felipe G. Foschiera
# URI 1288 - Canhão de Destruição
def knapsack(W, wg, val, n):
    if ks[n-1][W] != 0: return ks[n-1][W]
    if W == 0 or n == 0: return 0
    if wg[n-1] > W:
        result = knapsack(W, wg, val, n-1)
    else:
        result = max(
            val[n-1] + knapsack(W-wg[n-1], wg, val, n-1), knapsack(W, wg, val, n-1))
    ks[n-1][W] = result
    return result

T = int(input())
for _ in range(T):
    N = int(input())
    danos, pesos = [], []
    for _ in range(N):
        d, p = map(int, input().split())
        danos.append(d)
        pesos.append(p)
    K = int(input())
    R = int(input())
    ks = [[0 for _ in range(K+1)] for _ in range(N)]
    result = knapsack(K, pesos, danos, N)
    if result >= R:
        print("Missao completada com sucesso")
    else:
        print("Falha na missao")

File: URI Online Judge/Paradigmas/1288 - Canhao de Destruicao/canhao.cpp
#include <stdio.h>
#include <vector>
using namespace std;
typedef vector<int> vi;

int knapsack(int W, vi wt, vi v, int n){
    int ks[n+1][W+1];
    for(int i = 0; i <= n; i++){
        for(int w = 0; w <= W; w++){
            if(i == 0 || w == 0) ks[i][w] = 0;
            else if(wt[i-1] > w) ks[i][w] = ks[i-1][w];
            else ks[i][w] = max(ks[i-1][w], v[i-1] + ks[i-1][w-wt[i-1]]);
        }
    }
    return ks[n][W];
}


int main(){
    int T;
    scanf("%d", &T);
    while(T--){
        int N;
        scanf("%d", &N);
        vi pesos, valores;
        int v, p;
        for(int i = 0; i < N; i++){
            scanf("%d %d", &v, &p);
            valores.push_back(v);
            pesos.push_back(p);
            
        }
        int K, R; 
        scanf("%d %d", &K, &R);
        int ans = knapsack(K, pesos, valores, N);
        printf(ans >= R ? "Missao completada com sucesso\n" : "Falha na missao\n");
    }
    return 0;   
}